## MLP to classify images from the MNIST dataset.the model is designed to recognize handwritten digits ,experimenting with hidden layers and backpropagation to improve results.
## Ibrahim Mohamed Shedoh

In [3]:
!pip install tensorflow

Defaulting to user installation because normal site-packages is not writeable
  Using cached tensorflow-2.18.0-cp312-cp312-win_amd64.whl.metadata (3.3 kB)
  Using cached tensorflow_intel-2.18.0-cp312-cp312-win_amd64.whl.metadata (4.9 kB)
  Using cached absl_py-2.1.0-py3-none-any.whl.metadata (2.3 kB)
  Using cached astunparse-1.6.3-py2.py3-none-any.whl.metadata (4.4 kB)
  Using cached flatbuffers-24.3.25-py2.py3-none-any.whl.metadata (850 bytes)
  Using cached gast-0.6.0-py3-none-any.whl.metadata (1.3 kB)
  Using cached google_pasta-0.2.0-py3-none-any.whl.metadata (814 bytes)
  Using cached libclang-18.1.1-py2.py3-none-win_amd64.whl.metadata (5.3 kB)
  Using cached opt_einsum-3.4.0-py3-none-any.whl.metadata (6.3 kB)
  Using cached termcolor-2.5.0-py3-none-any.whl.metadata (6.1 kB)
  Using cached grpcio-1.67.1-cp312-cp312-win_amd64.whl.metadata (4.0 kB)
  Using cached tensorboard-2.18.0-py3-none-any.whl.metadata (1.6 kB)
  Using cached keras-3.6.0-py3-none-any.whl.metadata (5.8 kB)
  Us

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [5]:
import tensorflow as tf
print(tf.__version__)

2.18.0


In [6]:
#Import Required Libraries

import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

In [7]:
#Prepare the Data:

# Example: Load the MNIST dataset
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

# Normalize the images to [0, 1] range
x_train = x_train.astype("float32") / 255
x_test = x_test.astype("float32") / 255

# Flatten the images (28x28 to 784)
x_train = x_train.reshape((x_train.shape[0], -1))
x_test = x_test.reshape((x_test.shape[0], -1))

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 9s 1us/step


In [8]:
#Define the MLP Model:
#Experiment with different numbers of hidden layers and neurons to see the impact on performance.

model = keras.Sequential([
    layers.Dense(128, activation='relu', input_shape=(784,)),  # Input layer
    layers.Dense(64, activation='relu'),                       # Hidden layer 1
    layers.Dense(32, activation='relu'),                       # Hidden layer 2
    layers.Dense(10, activation='softmax')                     # Output layer (10 classes for MNIST)
])

C:\Users\ibrez\AppData\Roaming\Python\Python312\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [9]:
#Compile the Model:
#Choose an appropriate loss function and optimizer.

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [10]:
#Train the Model:
#Train using backpropagation.

model.fit(x_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

Epoch 1/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.8511 - loss: 0.5141 - val_accuracy: 0.9582 - val_loss: 0.1418
Epoch 2/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9612 - loss: 0.1278 - val_accuracy: 0.9647 - val_loss: 0.1238
Epoch 3/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9731 - loss: 0.0863 - val_accuracy: 0.9699 - val_loss: 0.1025
Epoch 4/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9799 - loss: 0.0642 - val_accuracy: 0.9702 - val_loss: 0.1080
Epoch 5/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9846 - loss: 0.0494 - val_accuracy: 0.9708 - val_loss: 0.1029
Epoch 6/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9876 - loss: 0.0398 - val_accuracy: 0.9714 - val_loss: 0.1009
Epoch 7/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9896 - loss: 0.0320 - val_accuracy: 0.9707 - val_loss: 0.1129
Epoch 8/10
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.9896 - loss: 0.0311 - 

In [11]:
#Evaluate the Model:
#Test its performance on unseen data.

test_loss, test_acc = model.evaluate(x_test, y_test)
print(f"Test accuracy: {test_acc:.2f}")

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.9643 - loss: 0.1589
Test accuracy: 0.97


In [12]:
#Analyze the Results:
from sklearn.metrics import confusion_matrix, classification_report
import numpy as np

y_pred = np.argmax(model.predict(x_test), axis=1)
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step
[[ 974    1    1    0    0    0    2    1    1    0]
 [   0 1115    3    1    0    1    7    1    7    0]
 [   5    0 1009    6    0    1    5    4    2    0]
 [   2    0    4  955    0   43    0    2    4    0]
 [   2    0    7    2  930    2   14    0    2   23]
 [   4    0    0    3    1  868    7    0    7    2]
 [   2    2    0    0    1    4  949    0    0    0]
 [   0    1   14    5    0    0    0 1002    3    3]
 [   7    0    6   11    0    6    9    3  932    0]
 [   3    2    0    9    2   16    2    5    6  964]]
              precision    recall  f1-score   support

           0       0.97      0.99      0.98       980
           1       0.99      0.98      0.99      1135
           2       0.97      0.98      0.97      1032
           3       0.96      0.95      0.95      1010
           4       1.00      0.95      0.97       982
           5       0.92      0.97      0.95       892
           6       0.95      0.99      0.97      

The analyzed results include a confusion matrix and classification report for your model's performance

1. Confusion Matrix:
The confusion matrix is a 10x10 table representing the true classes (rows) and predicted classes (columns) for each digit (0-9). Each element at position (i, j) indicates how many instances of class i were classified as class j. Ideally, values should be concentrated along the diagonal, where predictions match the true labels.

Observations:

Diagonal values (e.g., 974, 1115, 1009, etc.) show correct predictions for each digit.
Off-diagonal values (e.g., 3 in row 1, column 8 or 43 in row 3, column 5) show misclassifications where the model predicted a different digit than the actual one.

2. Classification Report:
The classification report provides more detailed metrics for each class:

Precision: The proportion of true positive predictions among all positive predictions for each class. High precision indicates a low false positive rate.
Recall: The proportion of true positive predictions among all actual positive instances for each class. High recall indicates a low false negative rate.
F1-score: The harmonic mean of precision and recall, balancing the two metrics.
Support: The number of true instances for each class.
Key Insights:

Class '1' has the highest precision (0.99), indicating that predictions for digit '1' are highly accurate when they are positive.
Class '5' has a slightly lower precision (0.92), suggesting that it is sometimes confused with other digits.
Recall for '3' (0.95) is lower due to some misclassifications, potentially with digit '5' as indicated by 43 misclassified instances.
Overall accuracy is 97%, indicating excellent performance but with some room for improvement.
Macro Average vs. Weighted Average:
Macro avg: The unweighted mean of precision, recall, and F1-score across all classes, treating each class equally.
Weighted avg: Takes into account the number of instances in each class, giving more importance to classes with more samples. This helps reflect the overall model performance considering class imbalances.

Conclusion:
Strengths: The model has high precision and recall across most classes, indicating strong overall performance.
Weaknesses: Some specific digits, such as '3' and '5', have relatively more misclassifications and could benefit from further optimization (e.g., more data, improved data augmentation, or architecture adjustments).
Next Steps: Perform error analysis on misclassified instances to understand patterns or potential reasons (e.g., overlapping features between '3' and '5') and consider tuning the model or retraining with additional data.